optd-cost-model/src/cost/filter/controller.rs

```diff
@@ -13,73 +13,78 @@ use crate::{
 impl<S: CostModelStorageLayer> CostModelImpl<S> {
     // TODO: is it a good design to pass table_id here? I think it needs to be refactored.
     // Consider to remove table_id.
-    pub fn get_filter_row_cnt(
+    pub async fn get_filter_row_cnt(
         &self,
         child_row_cnt: EstimatedStatistic,
         cond: ArcPredicateNode,
     ) -> CostModelResult<EstimatedStatistic> {
-        let selectivity = { self.get_filter_selectivity(cond)? };
+        let selectivity = { self.get_filter_selectivity(cond).await? };
         Ok(
             EstimatedStatistic((child_row_cnt.0 as f64 * selectivity) as u64)
                 .max(EstimatedStatistic(1)),
         )
     }
 
-    pub fn get_filter_selectivity(&self, expr_tree: ArcPredicateNode) -> CostModelResult<f64> {
-        match &expr_tree.typ {
-            PredicateType::Constant(_) => Ok(Self::get_constant_selectivity(expr_tree)),
-            PredicateType::AttributeRef => unimplemented!("check bool type or else panic"),
-            PredicateType::UnOp(un_op_typ) => {
-                assert!(expr_tree.children.len() == 1);
-                let child = expr_tree.child(0);
-                match un_op_typ {
-                    // not doesn't care about nulls so there's no complex logic. it just reverses
-                    // the selectivity for instance, != _will not_ include nulls
-                    // but "NOT ==" _will_ include nulls
-                    UnOpType::Not => Ok(1.0 - self.get_filter_selectivity(child)?),
-                    UnOpType::Neg => panic!(
-                        "the selectivity of operations that return numerical values is undefined"
-                    ),
+    pub async fn get_filter_selectivity(
+        &self,
+        expr_tree: ArcPredicateNode,
+    ) -> CostModelResult<f64> {
+        Box::pin(async move {
```

Comment:
Why would we need this here?

Reply:
Rust says if you want to use an async method recursively, you need this :(